### DISCUSSION TITLE
How would you implement "long" Keyboard shortcut sequences in Avalonia?

### DISCUSSION THREAD

**Discussion by uxmal** [AvaloniaUI/Avalonia#17476] [Q&A]
(Asked on StackOverflow,  https://stackoverflow.com/questions/79177890, but thought I'd ask here too).

Some applications, like Visual Studio, support long keyboard shortcut sequences. For instance, the command `Toggle bookmark` is bound to the sequence `Ctrl+K Ctrl+K`. When writing an Avalonia application, how would you approach implementing this keyboard sequence?

Since `KeyGesture` only supports a single (`modifier` + `Key`) combination, it seems I'd have to build an intermediate handler that reacts to the initial `Ctrl+K` (maybe displaying a status bar message like `"Waiting for next key press..."`). The next `Ctrl+K` would then have to be handled by the intermediate handler, which would dispatch to an actual `ToggleBookmark` command -- or handle the `ESC` key to get out of the modal state introduced by this mechanism.

The kicker is that once the handler has received `Ctrl+K`, I have to rewire the window so that _all_ keyboard events are routed to intermediate handler, and after the second event has been handled, the rewiring has to be restored to its original state. It's unclear how I'd do this, and whether it is the right approach. Some way to push a `HotKeyManager` on a stack of hotkey managers maybe?

Is this The Way or is there a more appropriate approach?

**Comment by timunie** [comment]
I think a private  List of Keys could be an option? If user released ctrl, clear that list. Whenever a key is pressed, evaluate the input

**Comment by StefanKoell** [comment]
I think your best bet would be to handle keyboard shortcuts yourself on a much lower level using something like https://github.com/TolikPylypchuk/SharpHook

You have to build some infrastructure around it to make it work but my initial tests were promising. 
The basic idea is to have some sort of command registry where all your commands are listed and also know about the keyboard shortcut they listens to.
When you register a shortcut, lookup if this is a single shortcut command and execute it immediately.
If the shortcut is only the first part of a "long" shortcut, remember it and wait some time for another shortcut. Once you receive the second shortcut, lookup in the command registry which shortcut matches and execute it.

You can do also some visual indication in a status bar like Jetbrains is doing. Depending on the complexity of your app you may also need to consider different contexts like current selection or so.
